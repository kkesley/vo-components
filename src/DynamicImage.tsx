import * as React from 'react'
import { style } from 'typestyle'

const styles = {
  image: style({
    objectFit: 'cover',
  }),
}

export type ImageSize = 1920 | 1280 | 1080 | 720 | 480 | 320 | 120 | 50

export interface ResponsiveConfig {
  maxScreenWidth: number
  imageWidth: ImageSize
}

export interface DynamicImageURLArgs {
  src: string
  width: ImageSize
}

export const getDynamicImageURL = (props: DynamicImageURLArgs) => {
  const { src, width } = props
  return src.replace('{dimension}', width.toString())
}

export interface DynamicImageProps extends DynamicImageURLArgs {
  alt?: string
  srcSet?: ResponsiveConfig[]
}

export default function DynamicImage(props: DynamicImageProps) {
  const { alt, src, width, srcSet } = props
  return (
    <picture>
      {(srcSet || []).map((config, index) => (
        <source
          key={`img-${alt}-${index}`}
          media={`(max-width: ${config.maxScreenWidth}px)`}
          srcSet={getDynamicImageURL({ src, width: config.imageWidth })}
        />
      ))}
      <img
        className={styles.image}
        alt={alt}
        src={getDynamicImageURL({ src, width })}
      />
    </picture>
  )
}
