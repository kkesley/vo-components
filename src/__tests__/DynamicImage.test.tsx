import React from 'react'
import renderer from 'react-test-renderer'
import DynamicImage, { getDynamicImageURL } from '../DynamicImage'

describe('DynamicImage', () => {
  const src = 'https://cdn.test.com/image/{dimension}/abc.jpg'
  const testCases = [
    {
      width: 1024,
      result: 'https://cdn.test.com/image/1024/abc.jpg',
    },
    {
      width: 720,
      result: 'https://cdn.test.com/image/720/abc.jpg',
    },
    {
      width: 999,
      result: 'https://cdn.test.com/image/999/abc.jpg',
    },
  ]
  testCases.forEach(({ width, result }) => {
    it(`returns the correct url for width: ${width}`, () => {
      expect(getDynamicImageURL({ src, width })).toEqual(result)
    })
    it(`has the correct src for width: ${width}`, () => {
      const image = renderer.create(<DynamicImage src={src} width={width} />)
        .root
      expect(image.findByType('img').props.src).toEqual(result)
    })
  })
  it('correctly passes the `alt` attribute to img', () => {
    const alt = 'hello'
    const image = renderer.create(
      <DynamicImage src="test.jpg" width={1024} alt={alt} />
    ).root
    expect(image.findByType('img').props.alt).toEqual(alt)
  })
})
