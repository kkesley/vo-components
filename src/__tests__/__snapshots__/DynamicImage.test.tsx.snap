// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicImage correctly creates source(s) when srcSet is configured 1`] = `
<picture>
  <source
    media="(max-width: 1920px)"
    srcSet="1280/test.jpg"
  />
  <source
    media="(max-width: 720px)"
    srcSet="480/test.jpg"
  />
  <img
    className="f5u4fpt"
    src="1920/test.jpg"
  />
</picture>
`;
